content/en/docs/usage/gateway-api.md

```diff
@@ -0,0 +1,347 @@
+---
+title: "Securing Gateway Resources"
+linkTitle: "Securing Gateway Resources"
+weight: 90
+type: "docs"
+---
+
+Since 1.5, cert-manager supports requesting TLS certificates using annotations
+on Gateway resources. This works similarly as to what you can do with
+annotations on the Ingress resource, as described on the page [Securing Ingress
+Resources](/docs/usage/ingress/).
+
+The Gateway resource is part of the [Gateway API][gwapi], a set of CRDs that can
+be installed to your Kubernetes cluster and that aim at offering a richer
+alternative to the Ingress abstraction.
+
+[gwapi]: https://gateway-api.sigs.k8s.io
+
+The Gateway resource holds the TLS configuration, as illustrated in the
+following diagram (source: https://gateway-api.sigs.k8s.io):
+
+![Gateway vs. HTTPRoute](/images/gateway-roles.png)
+
+Note that cert-manager only supports setting up the TLS configuration on the
+Gateway resource when the Gateway is configured to terminate the TLS connection.
+There is still uncertainty around how the HTTPRoute TLS configuration should be
+used (see [the discussion about TLS interaction between HTTPRoute and
+Gateway][gateway-api#577]) and no existing implementation supports the TLS
+pass-through mode in which the HTTPRoute contains the TLS configuration and
+terminates the connection (see the [feature request for supporting `spec.tls` in
+HTTPRoute for Istio][istio#31747]).
+
+[istio#31747]: https://github.com/istio/istio/issues/31747
+[gateway-api#577]: https://github.com/kubernetes-sigs/gateway-api/issues/577
+
+Before enabling the Gateway support, you will have to install the Gateway API
+CRDs:
+
+```sh
+kubectl kustomize "github.com/kubernetes-sigs/gateway-api/config/crd?ref=v0.3.0" | kubectl apply -f -
+```
+
+You then need to manually enable the Gateway support for Certificate using a
+flag:
+
+- If you are using Helm:
+
+  ```sh
+  helm upgrade --install cert-manager jetstack/cert-manager
+    --set "extraArgs={--controllers=*\,gateway-shim}"
+  ```
+
+- If you are using the raw cert-manager manifests, you can add the following
+  argument to the cert-manager controller Deployment's `args`:
+
+  ```yaml
+  args:
+    # ...
+    - --controllers=*,gateway-shim
+  ```
+
+You can also install the CRDs after installing cert-manager. cert-manager will
+pick up the Gateway CRD as soon as it is installed.
```

Comment:
Is this true? I would have thought the controller needs to be restarted. How are we doing this?

Reply:
Not sure anymore. At the moment, we are discovering whether the CRDs are present or not, but the watchers still keep watching. It goes like this (see [controller.go](https://github.com/jetstack/cert-manager/pull/4276/files)):

```go
	// check if the gateway API CRDs are available
	var gatewayAvailable bool
	d := cl.Discovery()
	resources, err := d.ServerResourcesForGroupVersion(gwapi.GroupVersion.String())
	switch {
	case apierrors.IsNotFound(err):
		gatewayAvailable = false
		log.Info("the Gateway API CRDs do not seem to be present, cert-manager will keep retrying watching for them")
	case err != nil:
		return nil, nil, fmt.Errorf("while checking if the Gateway API CRD is installed: %s", err.Error())
	case len(resources.APIResources) == 0:
		gatewayAvailable = false
		log.Info("the Gateway API CRDs do not seem to be present, cert-manager will keep retrying watching for them")
	default:
		gatewayAvailable = true
	}
```

I'll remove this line since it feels quite tricky to get this "crd-hot-reloading" mechanism right, and it might not be worth it.

cc @irbekrm @jakexks

Reply:
I see, taking a look at the code it looks like “ cert-manager will keep retrying watching for them” is incorrect. Unless I’ve missed some mechanism, we would need to setup some kind of loop to keep trying to see if they’re available.

IMO, we should _not_ do this. If a user wants to use this API, it is expected behaviour that we would error exit the program if they are not available. “Silently” continuing is bad behaviour I think.

This is inline with how all other operators+CRDs behave, especially in cert-manager.

Reply:
From these past conversations:

1. https://github.com/jetstack/cert-manager/pull/4158#discussion_r669380637
2. https://github.com/jetstack/cert-manager/pull/4276#discussion_r679364495
3. https://github.com/jetstack/cert-manager/pull/4276#discussion_r680174007

There are two separate concerns that we have been trying to figure out:

### Concern 1: how do we turn the feature on?

Should we discover the Gateway API, or should we use gate the feature behind flags such as `--controllers=*,gateway-shim` and `--controllers=*,gateway-http01`?

→ At first, we went with a flag (`--controllers=*,gateway-shim`). @jakexks then went instead with discovering after the discussion in https://github.com/jetstack/cert-manager/pull/4276#discussion_r679364495.


### Concern 2: how do we handle the fact that Gateway API is not a core API?

When cert-manager is installed before the CRDs are installed, should we retry?

→ In https://github.com/jetstack/cert-manager/pull/4158#discussion_r669380637, @irbekrm made the point that since Gateway API is not part of core Kubernetes we shouldn't crash: 

> Can we avoid crashing here? Because Gateway CRDs are not part of the core API, I imagine they could not yet be installed at this point, depending on users' workflow and it may be a better user experience if the pod didn't crash upfront if they forget to install the gateway controller first. Or, would this just cause issues down the line? Alternatively, if we keep this crash, perhaps the error message could be like gateway-shim controller is enabled, but couldn't discover ... to make debugging easier etc.

Reply:
This all makes sense to me!

My concern, as @irbekrm mentioned, is how can we avoid the case where a user installs things in the wrong order, and then is confused as to why nothing is working. I agree, this _will_ become GitHub issues and slack messages down the road.

At least a pod crashing is a clear indicator that something is wrong and a user won’t miss anything.

Reply:
Here is how it works currently:

**Gateway shim:**

- To be able to use the gateway shim feature, users have to pass `--controllers=*,gateway-shim`.
- When cert-manager is started with the gateway shim feature turned on but the Gateway API CRDs are not yet installed, the user will be warned that the CRD should be installed.
- The user can install the CRD after starting cert-manager, and cert-manager will pick it up and the gateway shim will start working.

**Gateway solver:**

- To be able to use the gateway solver, users need to install the Gateway API CRDs,
- When cert-manager is started without the CRDs installed, cert-manager says that the Gateway HTTP-01 solver support is disabled due to the CRDs no being installed,
- In order to install the CRDs after cert-manager is running, you have to restart cert-manager to make the gateway solver feature start working.

Reply:
@wallrj made a good point during this morning's standup: why not turn on/off both features simultaneously with a feature gate, e.g.:

```
--feature-gates=GatewayAPI=true
```

Users would still have to restart the controllers if the CRDs are installed when cert-manager is already running to start the "Gateway solver" feature (the "Gateway shim" feature would pick up the CRDs when they are installed).

That said, we still decided to go with the current implementation (described [above](https://github.com/cert-manager/website/pull/640#discussion_r682481309)) for releasing 1.5.0-beta.0.